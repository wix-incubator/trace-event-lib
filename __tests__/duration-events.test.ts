import { trace } from './setup-snapshot-testing';

test('single process', async () => {
  // Instant event
  trace.instant({ cat: 'config', name: 'Resolve config', ts: 0, args: { testTimeout: 60_000 } });

  // Duration events: separate calls
  trace.begin({ cat: 'device', name: 'Boot a device', ts: 0 });
  trace.end({ ts: 1e6 });

  // Duration events: chainable call
  trace.begin({ cat: 'test', name: 'Run a test suite', ts: 1e6 }).end({ ts: 1e7 });

  // Complete event
  trace.complete({ cat: 'device', name: 'Shutdown the device', ts: 1e7, dur: 3e6 });
});

test('multiple processes', async () => {
  // Metadata events: process_name - simplified
  trace.process_name('cli', 1);
  // Metadata events: process_name - explicit
  trace.metadata({ pid: 2, name: 'process_name', args: { name: 'jest' } });

  // Process 1
  trace.complete({ cat: 'spawn', name: 'jest --runInBand', ts: 0, dur: 2.5e6 });
  // Process 2
  trace.complete({ pid: 2, cat: 'run', name: '01.test.js', ts: 0.25e6, dur: 1e6 });
  trace.complete({ pid: 2, cat: 'run', name: '02.test.js', ts: 1.25e6, dur: 1e6 });
});

test('multiple threads', async () => {
  // Simple thread (1)
  trace.begin({ tid: 1, cat: 'device', name: 'boot', ts: 0 }).end({ ts: 1e6 });
  trace.complete({ tid: 1, cat: 'device', name: 'install app', ts: 1.25e6, dur: 0.75e6 });
  trace.complete({ tid: 1, cat: 'device', name: 'launch app', ts: 4e6, dur: 0.25e6 });

  // Stacked thread (0)
  trace.begin({
    tid: 0,
    cat: 'lifecycle',
    name: 'Dashboard - should have a currency widget',
    ts: 3e6,
  });
  trace.begin({ tid: 0, cat: 'lifecycle', name: 'beforeEach', ts: 4e6 }).end({ ts: 5e6 });
  trace.begin({ tid: 0, cat: 'lifecycle', name: 'testFn', ts: 6e6 }).end({ ts: 8e6 });
  trace.begin({ tid: 0, cat: 'lifecycle', name: 'afterEach', ts: 9e6 }).end({ ts: 10e6 });
  trace.end({ tid: 0, ts: 11e6 });

  // Metadata events: thread_name - explicit
  trace.metadata({ tid: 0, name: 'thread_name', args: { name: 'Test lifecycle' } });
  // Metadata events: thread_name - implicit
  trace.thread_name('Device', 1);

  // Metadata events: thread_sort_index - explicit
  trace.metadata({ tid: 0, name: 'thread_sort_index', args: { sort_index: 2 } });
  // Metadata events: thread_sort_index - implicit
  trace.thread_sort_index(1, 1);
});
