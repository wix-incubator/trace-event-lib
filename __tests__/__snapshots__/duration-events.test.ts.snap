// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple processes 1`] = `
Array [
  Object {
    "args": Object {
      "name": "cli",
    },
    "name": "process_name",
    "ph": "M",
    "pid": 1,
    "tid": 0,
    "ts": 0,
  },
  Object {
    "args": Object {
      "name": "jest",
    },
    "name": "process_name",
    "ph": "M",
    "pid": 2,
    "tid": 0,
    "ts": 0,
  },
  Object {
    "cat": "spawn",
    "dur": 2500000,
    "name": "jest --runInBand",
    "ph": "X",
    "pid": 1,
    "tid": 0,
    "ts": 0,
  },
  Object {
    "cat": "run",
    "dur": 1000000,
    "name": "01.test.js",
    "ph": "X",
    "pid": 2,
    "tid": 0,
    "ts": 250000,
  },
  Object {
    "cat": "run",
    "dur": 1000000,
    "name": "02.test.js",
    "ph": "X",
    "pid": 2,
    "tid": 0,
    "ts": 1250000,
  },
]
`;

exports[`multiple threads 1`] = `
Array [
  Object {
    "cat": "device",
    "name": "boot",
    "ph": "B",
    "pid": 1,
    "tid": 1,
    "ts": 0,
  },
  Object {
    "ph": "E",
    "pid": 1,
    "tid": 1,
    "ts": 1000000,
  },
  Object {
    "cat": "device",
    "dur": 750000,
    "name": "install app",
    "ph": "X",
    "pid": 1,
    "tid": 1,
    "ts": 1250000,
  },
  Object {
    "cat": "device",
    "dur": 250000,
    "name": "launch app",
    "ph": "X",
    "pid": 1,
    "tid": 1,
    "ts": 4000000,
  },
  Object {
    "cat": "lifecycle",
    "name": "Dashboard - should have a currency widget",
    "ph": "B",
    "pid": 1,
    "tid": 0,
    "ts": 3000000,
  },
  Object {
    "cat": "lifecycle",
    "name": "beforeEach",
    "ph": "B",
    "pid": 1,
    "tid": 0,
    "ts": 4000000,
  },
  Object {
    "ph": "E",
    "pid": 1,
    "tid": 0,
    "ts": 5000000,
  },
  Object {
    "cat": "lifecycle",
    "name": "testFn",
    "ph": "B",
    "pid": 1,
    "tid": 0,
    "ts": 6000000,
  },
  Object {
    "ph": "E",
    "pid": 1,
    "tid": 0,
    "ts": 8000000,
  },
  Object {
    "cat": "lifecycle",
    "name": "afterEach",
    "ph": "B",
    "pid": 1,
    "tid": 0,
    "ts": 9000000,
  },
  Object {
    "ph": "E",
    "pid": 1,
    "tid": 0,
    "ts": 10000000,
  },
  Object {
    "ph": "E",
    "pid": 1,
    "tid": 0,
    "ts": 11000000,
  },
  Object {
    "args": Object {
      "name": "Test lifecycle",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1,
    "tid": 0,
    "ts": 0,
  },
  Object {
    "args": Object {
      "name": "Device",
    },
    "name": "thread_name",
    "ph": "M",
    "pid": 1,
    "tid": 1,
    "ts": 0,
  },
  Object {
    "args": Object {
      "sort_index": 2,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1,
    "tid": 0,
    "ts": 0,
  },
  Object {
    "args": Object {
      "sort_index": 1,
    },
    "name": "thread_sort_index",
    "ph": "M",
    "pid": 1,
    "tid": 1,
    "ts": 0,
  },
]
`;

exports[`single process 1`] = `
Array [
  Object {
    "args": Object {
      "testTimeout": 60000,
    },
    "cat": "config",
    "name": "Resolve config",
    "ph": "i",
    "pid": 1,
    "tid": 0,
    "ts": 0,
  },
  Object {
    "cat": "device",
    "name": "Boot a device",
    "ph": "B",
    "pid": 1,
    "tid": 0,
    "ts": 0,
  },
  Object {
    "ph": "E",
    "pid": 1,
    "tid": 0,
    "ts": 1000000,
  },
  Object {
    "cat": "test",
    "name": "Run a test suite",
    "ph": "B",
    "pid": 1,
    "tid": 0,
    "ts": 1000000,
  },
  Object {
    "ph": "E",
    "pid": 1,
    "tid": 0,
    "ts": 10000000,
  },
  Object {
    "cat": "device",
    "dur": 3000000,
    "name": "Shutdown the device",
    "ph": "X",
    "pid": 1,
    "tid": 0,
    "ts": 10000000,
  },
]
`;
